### PR TITLE
Move git-machete-specific logic away from gitCore into backend

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

File: gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreReflogEntry.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import lombok.ToString;
+import org.eclipse.jgit.lib.ReflogEntry;
+
+import com.virtuslab.gitcore.api.IGitCoreCommitHash;
+import com.virtuslab.gitcore.api.IGitCoreReflogEntry;
+
+@EqualsAndHashCode
+@Getter
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+@ToString
```
Comment (micpiotrowski) [inline gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreReflogEntry.java:17]: Can be simplified
```suggestion
@Data
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
```
Lombok is smart enough to make ctor private anyway.

Comment (micpiotrowski) [reply gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreReflogEntry.java:17]: The same probably in other places

File: gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/BaseGitCoreBranch.java [hunk 2]
```diff
@@ -1,147 +1,41 @@
...
 
-  @SuppressWarnings("regexp") // to allow `synchronized`
   @Override
-  public synchronized GitCoreCommit getPointedCommit() throws GitCoreException {
-    if (pointedCommit == null) {
-      pointedCommit = repo.revStringToGitCoreCommit(getFullName());
-    }
-    return pointedCommit;
+  public GitCoreCommit derivePointedCommit() throws GitCoreException {
+    return pointedCommit.get().getOrElseThrow(Function.identity());
```
[the remark this answers is not in the document]
Comment (PawelLipski) [reply gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/BaseGitCoreBranch.java:33]: Done

File: gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreReflogEntry.java (hunk 1 above)
Comment (PawelLipski) [reply gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreReflogEntry.java:17]: Done here and in `GitCorePersonIdentity`

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java [hunk 3]
```diff
@@ -62,295 +67,415 @@ public IGitMacheteRepository create(Path mainDirectoryPath, Path gitDirectoryPat
...
-    var rootBranchTries = branchLayout.getRootEntries()
-        .map(entry -> Try.of(() -> createGitMacheteRootBranch(gitCoreRepository, statusHookExecutor, entry)));
-    var rootBranches = Try.sequence(rootBranchTries).getOrElseThrow(GitMacheteException::castOrWrap);
+    var result = new Aux(gitCoreRepository, statusHookExecutor).createGitMacheteRepository(branchLayout);
+    LOG.withTimeElapsed().debug("Finished");
+    return result;
+  }
 
-    var branchByName = createBranchByNameMap(rootBranches);
+  private static class Aux {
```
Comment (mkondratek) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java:75]: :open_mouth:

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java [hunk 4]
```diff
@@ -62,295 +67,415 @@ public IGitMacheteRepository create(Path mainDirectoryPath, Path gitDirectoryPat
...
+      if (remoteBranchOption.isDefined()) {
+        IGitCoreRemoteBranch coreRemoteBranch = remoteBranchOption.get();
+        IGitCoreCommit coreRemoteBranchPointedCommit = Try.of(() -> coreRemoteBranch.derivePointedCommit())
+            .getOrElseThrow(e -> new GitMacheteException("Cannot get core remote branch pointed commit", e));
+        remoteBranch = new GitMacheteRemoteBranch(new GitMacheteCommit(coreRemoteBranchPointedCommit));
+      }
+      return remoteBranch;
+    }
+
+    private Option<IGitCoreCommit> deriveParentAgnosticForkPoint(IGitCoreLocalBranch branch) throws GitCoreException {
```
Comment (mkondratek) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java:181]: :xdddd:

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java [hunk 5]
```diff
@@ -89,8 +89,7 @@ public FetchResultImpl fetchDefaultRemote(Collection<GitRepository> repositories
             return false;
           }
 
-        }).map(repoAndRemote -> new RemoteRefCoordinates(repoAndRemote._1(), repoAndRemote._2(), /* refspec */ null))
-        .collect(List.collector());
```
Comment (mkondratek) [inline frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java:93]: ~Good point! Maybe some regexp for that?~
image

okay, sometimes it actually is needed... but can be managed like this from:
image
to:
image

File: gitCoreApi/src/main/java/com/virtuslab/gitcore/api/IGitCoreRepository.java [hunk 6]
```diff
@@ -1,24 +1,30 @@
...
-  Path getMainDirectoryPath();
+  List<IGitCoreRemoteBranch> deriveAllRemoteBranches() throws GitCoreException;
+
+  List<IGitCoreRemoteBranch> deriveRemoteBranchesForRemote(String remoteName) throws GitCoreException;
 
   boolean isAncestor(IGitCoreCommit presumedAncestor, IGitCoreCommit presumedDescendant) throws GitCoreException;
+
+  Option<IGitCoreCommit> findFirstAncestor(
+      IGitCoreCommit fromInclusive,
+      Predicate<IGitCoreCommitHash> predicate) throws GitCoreException;
```
Comment (mkondratek) [inline gitCoreApi/src/main/java/com/virtuslab/gitcore/api/IGitCoreRepository.java:27]: :1st_place_medal:

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java (hunk 5 above)
Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java:93]: :trollface:

Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java:93]: Done indeed (via `FenumChecker`) on PR #329 :bangbang:

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java (hunk 3 above)
Comment (PawelLipski) [reply backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java:75]: No better idea how to do it :/

File: frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java (hunk 5 above)
Comment (PawelLipski) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java:93]: See https://github.com/VirtusLab/git-machete-intellij-plugin/pull/329/commits/13058bb43398213101b1c7beb4197e7afb822789

Comment (mkondratek) [reply frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java:93]: cool! :rocket:

**Pull request merged**